src/main/java/com/lafaspot/pop/session/PopSession.java

```diff
@@ -116,10 +138,10 @@ protected void initChannel(final SocketChannel ch) throws Exception {
             throw new PopException(Type.CONNECT_FAILURE, e);
         }
 
-        stateRef.compareAndSet(State.NULL, State.COMMAND_SENT);
```

Comment:
do the check + set and fail also, move the check before actual network operation, same for all other cases.

Reply:
Moved the check+set and fail fast to the beginning.